### PR TITLE
Support nemotron-speech-streaming-en-0.6b

### LINKED ISSUE #3040: Nemotron
Would it be possible to port over the [Nemotron](https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b) models (or at least the streaming model)? I was trying something like this, and tried to base it on the scripts that are already in the repo, but I can't seem to get it work properly:

```python
#!/usr/bin/env python3
import os
os.environ["TORCH_ONNX_USE_NEW_EXPORTER"] = "0"
os.environ["PYTORCH_ONNX_DYNAMO_EXPORT"] = "0"

import nemo.collections.asr as nemo_asr
import onnx
import torch
from onnxruntime.quantization import QuantType, quantize_dynamic
from pathlib import Path

# Force legacy ONNX export
torch.onnx.dynamo_export = None

def check_and_fix_names(filename: str):
    """Check and ensure all inputs/outputs have non-empty names"""
    model = onnx.load(filename)
    
    print(f"\n=== Checking {filename} ===")
    
    # Check inputs
    print("Inputs:")
    for i, inp in enumerate(model.graph.input):
        if not inp.name or inp.name.strip() == "":
            print(f"  ⚠️  Input {i} has empty name!")
            # Fix: assign a name
            inp.name = f"input_{i}"
        else:
            print(f"  ✅ Input {i}: '{inp.name}'")
    
    # Check outputs
    print("Outputs:")
    for i, out in enumerate(model.graph.output):
        if not out.name or out.name.strip() == "":
            print(f"  ⚠️  Output {i} has empty name!")
            # Fix: assign a name
            out.name = f"output_{i}"
        else:
            print(f"  ✅ Output {i}: '{out.name}'")
    
    # Save fixed model
    onnx.save(model, filename)
    print(f"Saved fixed model to {filename}")

def add_meta_data(filename: str, meta_data: dict):
    model = onnx.load(filename)
    while len(model.metadata_props):
        model.metadata_props.pop()
    for key, value in meta_data.items():
        meta = model.metadata_props.add()
        meta.key = key
        meta.value = str(value)
    onnx.save(model, filename)

@torch.no_grad()
def main():
    out_dir = Path("nemotron-onnx")
    out_dir.mkdir(exist_ok=True)
    os.chdir(out_dir)
    
    asr_model = nemo_asr.models.ASRModel.from_pretrained("nvidia/nemotron-speech-streaming-en-0.6b")
    asr_model.eval()

    # Save tokens
    with open("tokens.txt", "w", encoding="utf-8") as f:
        for i, s in enumerate(asr_model.joint.vocabulary):
            f.write(f"{s} {i}\n")
        f.write(f"<blk> {i+1}\n")
    print("Saved tokens.txt")

    # Get streaming config
    cfg = asr_model.encoder.streaming_cfg
    chunk_size = cfg.chunk_size[1] if isinstance(cfg.chunk_size, list) else cfg.chunk_size
    pre_encode_cache_size = cfg.pre_encode_cache_size[1] if isinstance(cfg.pre_encode_cache_size, list) else cfg.pre_encode_cache_size
    window_size = chunk_size + pre_encode_cache_size

    print(f"chunk_size: {chunk_size}")
    print(f"pre_encode_cache_size: {pre_encode_cache_size}")
    print(f"window_size: {window_size}")

    # Cache dimensions
    cache_last_channel_dim1 = len(asr_model.encoder.layers)
    cache_last_channel_dim2 = cfg.last_channel_cache_size
    cache_last_channel_dim3 = asr_model.encoder.d_model
    cache_last_time_dim1 = len(asr_model.encoder.layers)
    cache_last_time_dim2 = asr_model.encoder.d_model
    cache_last_time_dim3 = asr_model.encoder.conv_context_size[0]

    # Export ONNX models
    asr_model.encoder.export("encoder.onnx", check_trace=False)
    asr_model.decoder.export("decoder.onnx", check_trace=False)
    asr_model.joint.export("joiner.onnx", check_trace=False)

    # Check and fix names in original models
    print("\n=== Checking original models ===")
    check_and_fix_names("encoder.onnx")
    check_and_fix_names("decoder.onnx")
    check_and_fix_names("joiner.onnx")

    os.system("ls -lh *.onnx")

    normalize_type = asr_model.cfg.preprocessor.normalize
    if normalize_type == "NA":
        normalize_type = ""

    meta_data = {
        "vocab_size": asr_model.decoder.vocab_size,
        "window_size": window_size,
        "chunk_shift": chunk_size,
        "normalize_type": normalize_type,
        "cache_last_channel_dim1": cache_last_channel_dim1,
        "cache_last_channel_dim2": cache_last_channel_dim2,
        "cache_last_channel_dim3": cache_last_channel_dim3,
        "cache_last_time_dim1": cache_last_time_dim1,
        "cache_last_time_dim2": cache_last_time_dim2,
        "cache_last_time_dim3": cache_last_time_dim3,
        "pred_rnn_layers": asr_model.decoder.pred_rnn_layers,
        "pred_hidden": asr_model.decoder.pred_hidden,
        "subsampling_factor": 8,
        "model_type": "EncDecRNNTBPEModel",
        "version": "1",
        "model_author": "NVIDIA",
        "url": "https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b",
    }

    # Quantize
    print("\n=== Quantizing models ===")
    for m in ["encoder", "decoder", "joiner"]:
        try:
            quantize_dynamic(
                model_input=f"{m}.onnx",
                model_output=f"{m}.int8.onnx",
                weight_type=QuantType.QUInt8 if m == "encoder" else QuantType.QInt8,
            )
            print(f"Successfully quantized {m}.onnx")
            
            # CRITICAL: Check and fix names after quantization!
            check_and_fix_names(f"{m}.int8.onnx")
            
        except Exception as e:
            print(f"Warning: Failed to quantize {m}.onnx: {e}")
    
    # Add metadata to both versions
    add_meta_data("encoder.onnx", meta_data)
    if os.path.exists("encoder.int8.onnx"):
        add_meta_data("encoder.int8.onnx", meta_data)
    
    os.system("ls -lh *.onnx")
    print("\nmeta_data:", meta_data)
    print("\n✅ Done! Check the input/output names above.")

if __name__ == "__main__":
    main()

```

### PR DESCRIPTION
**Pull request by csukuangfj**
Fixes #3040 

CC @Dokotela

# Usage

1. Build the code from source
2. Download a model
```
wget https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14.tar.bz2

tar xvf sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14.tar.bz2
rm sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14.tar.bz2
```

3. Run it
```
  ./build/bin/sherpa-onnx \
    --encoder=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/encoder.int8.onnx \
    --decoder=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/decoder.int8.onnx \
    --joiner=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/joiner.int8.onnx \
    --tokens=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/tokens.txt \
    ./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/test_wavs/0.wav
```

The output logs are
```
/Users/fangjun/open-source/sherpa-onnx/sherpa-onnx/csrc/parse-options.cc:Read:373 ./build/bin/sherpa-onnx --encoder=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/encoder.int8.onnx --decoder=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/decoder.int8.onnx --joiner=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/joiner.int8.onnx --tokens=./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/tokens.txt ./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/test_wavs/0.wav

OnlineRecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=16000, feature_dim=80, low_freq=20, high_freq=-400, dither=0, normalize_samples=True, snip_edges=False), model_config=OnlineModelConfig(transducer=OnlineTransducerModelConfig(encoder="./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/encoder.int8.onnx", decoder="./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/decoder.int8.onnx", joiner="./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/joiner.int8.onnx"), paraformer=OnlineParaformerModelConfig(encoder="", decoder=""), wenet_ctc=OnlineWenetCtcModelConfig(model="", chunk_size=16, num_left_chunks=4), zipformer2_ctc=OnlineZipformer2CtcModelConfig(model=""), nemo_ctc=OnlineNeMoCtcModelConfig(model=""), t_one_ctc=OnlineToneCtcModelConfig(model=""), provider_config=ProviderConfig(device=0, provider="cpu", cuda_config=CudaConfig(cudnn_conv_algo_search=1), trt_config=TensorrtConfig(trt_max_workspace_size=2147483647, trt_max_partition_iterations=10, trt_min_subgraph_size=5, trt_fp16_enable="True", trt_detailed_build_log="False", trt_engine_cache_enable="True", trt_engine_cache_path=".", trt_timing_cache_enable="True", trt_timing_cache_path=".",trt_dump_subgraphs="False" )), tokens="./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/tokens.txt", num_threads=1, warm_up=0, debug=False, model_type="", modeling_unit="cjkchar", bpe_vocab=""), lm_config=OnlineLMConfig(model="", scale=0.5, lodr_scale=0.01, lodr_fst="", lodr_backoff_id=-1, shallow_fusion=True), endpoint_config=EndpointConfig(rule1=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=2.4, min_utterance_length=0), rule2=EndpointRule(must_contain_nonsilence=True, min_trailing_silence=1.2, min_utterance_length=0), rule3=EndpointRule(must_contain_nonsilence=False, min_trailing_silence=0, min_utterance_length=20)), ctc_fst_decoder_config=OnlineCtcFstDecoderConfig(graph="", max_active=3000), enable_endpoint=True, max_active_paths=4, hotwords_score=1.5, hotwords_file="", decoding_method="greedy_search", blank_penalty=0, temperature_scale=2, rule_fsts="", rule_fars="", reset_encoder=False, hr=HomophoneReplacerConfig(lexicon="", rule_fsts=""))
Start to create recognizer
Recognizer created in 0.75036 s
./sherpa-onnx-nemotron-speech-streaming-en-0.6b-int8-2026-01-14/test_wavs/0.wav
Number of threads: 1, Elapsed seconds: 0.63, Audio duration (s): 6.6, Real time factor (RTF) = 0.63/6.6 = 0.095
 after early nightfall the yellow lamps would light up here and there, squid quarter of the brothels
{ "text": " after early nightfall the yellow lamps would light up here and there, squid quarter of the brothels", "tokens": [" a", "fter", " ear", "ly", " n", "ight", "f", "all", " the", " y", "e", "llow", " l", "am", "ps", " w", "ould", " l", "ight", " up", " he", "re", " and", " the", "re", ",", " s", "qu", "id", " qu", "ar", "ter", " of", " the", " br", "oth", "els"], "timestamps": [0.40, 0.48, 0.56, 0.64, 1.12, 1.20, 1.36, 1.44, 1.68, 1.92, 2.00, 2.08, 2.24, 2.32, 2.40, 2.48, 2.56, 2.64, 2.72, 2.80, 3.36, 3.44, 3.52, 3.60, 3.68, 4.48, 4.56, 4.64, 4.72, 4.80, 4.88, 4.96, 5.04, 5.12, 5.60, 5.68, 5.92], "ys_probs": [], "lm_probs": [], "context_scores": [], "segment": 0, "words": [], "start_time": 0.00, "is_final": false, "is_eof": false}
```

## Summary by CodeRabbit

* **New Features**
  * Added support for exporting NeMo speech-streaming models to ONNX and publishing them to external model hubs.
  * Exposed a FeatureDim() API to retrieve model feature dimensionality.

* **Bug Fixes**
  * Removed unnecessary left-padding in audio input processing for streaming recognition.

* **Chores**
  * Added an automated workflow to export, package, and distribute NeMo ASR models.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by fosple** [comment]
@csukuangfj Thanks for converting the nemo model to onnx so it can be used with sherpa-onnx. Highly appreciated ❤️ 

I have a question: The original model needs certain fixed chunk sizes to work correctly (see below). How do I set those in sherpa-onnx? Or do I have to re-export the model using your [conversion script export_onnx.py](https://github.com/csukuangfj/sherpa-onnx/blob/master/scripts/nemo/nemotron-speech-streaming-en-0.6b/export_onnx.py) for a different chunk size? If I have to re-export how to set different chunk sizes?

Original model card <https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b>:
```
Latency is defined by the att_context_size param, where att_context_size = {num_frames_left_context, num_frame_right_context}, all measured in 80ms frames:

[70, 0]: Chunk size = 1 (1 × 80ms = 0.08s)
[70, 1]: Chunk size = 2 (2 × 80ms = 0.16s)
[70, 6]: Chunk size = 7 (7 × 80ms = 0.56s)
[70, 13]: Chunk size = 14 (14 × 80ms = 1.12s)
```

Thanks :)

**Comment by csukuangfj** [comment]
@fosple 

Thanks for your interest!

You need to re-export the model by following the changes below.

Change
https://github.com/k2-fsa/sherpa-onnx/blob/8dc6244e9d49fb17d410d3eab2244755b412386b/scripts/nemo/nemotron-speech-streaming-en-0.6b/export_onnx.py#L60
to

```
 asr_model.eval() 

# chunk size = 1 (80 ms)
asr_model.encoder.set_default_att_context_size([70, 0])
```

or

```
 asr_model.eval() 

# chunk size = 2 (160 ms)
asr_model.encoder.set_default_att_context_size([70, 1])
```

or

```
 asr_model.eval() 

# chunk size = 7 (560 ms)
asr_model.encoder.set_default_att_context_size([70, 6])
```

or

```
 asr_model.eval() 

# chunk size = 14 (1.12 ms), this is the default one if you don't do anything
asr_model.encoder.set_default_att_context_size([70, 13])
```

You can use
```
print(asr_model.encoder.att_context_size, asr_model.encoder.att_context_size)
print(asr_model.encoder.streaming_cfg)
```
to check that.

<table>
<tr>
<th>Config</th>
<th>Output</th>
</tr>

<tr>
<td><code>[70, 0]</code></td>
<td>
<pre lang="python">
[70, 0] [70, 0]
CacheAwareStreamingConfig(
chunk_size=[1, 8], shift_size=[1, 8], 
cache_drop_size=0, last_channel_cache_size=70, 
valid_out_len=1, pre_encode_cache_size=[0, 9], 
drop_extra_pre_encoded=2, last_channel_num=0, 
last_time_num=0)
</pre>
</td>
</tr>

<tr>
<td><code>[70, 1]</code></td>
<td>
<pre lang="python">
[70, 1] [70, 1]
CacheAwareStreamingConfig(
chunk_size=[9, 16], shift_size=[9, 16], 
cache_drop_size=0, last_channel_cache_size=70, 
valid_out_len=2, pre_encode_cache_size=[0, 9], 
drop_extra_pre_encoded=2, last_channel_num=0, 
last_time_num=0)
</pre>
</td>
</tr>

<tr>
<td><code>[70, 6]</code></td>
<td>
<pre lang="python">
[70, 6] [70, 6]
CacheAwareStreamingConfig(
chunk_size=[49, 56], shift_size=[49, 56], 
cache_drop_size=0, last_channel_cache_size=70, 
valid_out_len=7, pre_encode_cache_size=[0, 9], 
drop_extra_pre_encoded=2, last_channel_num=0, 
last_time_num=0)
</pre>
</td>
</tr>

<tr>
<td><code>[70, 13]</code></td>
<td>
<pre lang="python">
[70, 13] [70, 13]
CacheAwareStreamingConfig(
chunk_size=[105, 112], shift_size=[105, 112],
 cache_drop_size=0, last_channel_cache_size=70, 
valid_out_len=14, pre_encode_cache_size=[0, 9], 
drop_extra_pre_encoded=2, last_channel_num=0, 
last_time_num=0)
</pre>
</td>
</tr>
</table>

**Comment by fosple** [comment]
Nice! Thanks for the super fast answer :) Will try it out right away.

**Comment by csukuangfj** [comment]
Does it work?

**Comment by wangfeng35** [comment]
March 12, 2026: [nemotron-asr-streaming ](https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b/tree/main) was released with updated checkpoint (trained on larger corpora).

**Comment by csukuangfj** [comment]
> March 12, 2026: [nemotron-asr-streaming ](https://huggingface.co/nvidia/nemotron-speech-streaming-en-0.6b/tree/main) was released with updated checkpoint (trained on larger corpora).

Can you create a separate issue?
